src/smt-inout/chc_inv_in.cc

```diff
@@ -0,0 +1,929 @@
+    if (! IN(search_name, term_container)) {
+      auto local_var = get_a_new_local_var_name();
+      // here we need to put the variable
+      term_container.insert( std::make_pair( search_name, 
+        SmtTermInfoVerilog(
+          local_var+bitslice , 
+          var_type(var_type::tp::Bool, 1, ""), 
+          this ) ) );
+      // assign the current expression to a local var
+      // and apply to it
+      local_vars.insert( std::make_pair(
+        local_var, SmtTermInfoVerilog(
+          args[0]->_translate,
+          args[0]->_type,
+          args[0]->_context)));
+    }
+    return & ( term_container[search_name] );
+  } // else 
+  ILA_ASSERT(! S_IN(')', args[0]->_translate));
+
+  std::string vlg_expr = args[0]->_translate + bitslice;
+  std::string search_name = "##bool_"  + vlg_expr;
+  if (! IN(search_name, term_container)) {
+    // here we need to put the variable
+    term_container.insert( std::make_pair( search_name, 
+      SmtTermInfoVerilog(
+        vlg_expr , 
+        var_type(var_type::tp::Bool, 1, ""), 
+        this ) ) );
+  } // end of insert if not found
+  return & ( term_container[search_name] );
+} // bit2bool
+
+
+DEFINE_OPERATOR(repeat) {
+  ILA_ASSERT(false) << "Unimplemented";
+  return nullptr;
+}
+DEFINE_OPERATOR(zero_extend) {
+  ILA_ASSERT(idx.size() == 1);
+  ILA_ASSERT(args.size() == 1);
+  ILA_ASSERT(args[0]->_type._type == var_type::tp::BV);
+  ILA_ASSERT(false) << "Unimplemented";
+  return nullptr;
+  
+}
+DEFINE_OPERATOR(sign_extend) {
+  ILA_ASSERT(idx.size() == 1);
+  ILA_ASSERT(args.size() == 1);
+  ILA_ASSERT(args[0]->_type._type == var_type::tp::BV);
+  ILA_ASSERT(false) << "Unimplemented";
+  return nullptr;
+  
+}
+DEFINE_OPERATOR(rotate_left) {
+  ILA_ASSERT(false) << "Unimplemented";
+  return nullptr;
+}
+DEFINE_OPERATOR(rotate_right) {
+  ILA_ASSERT(false) << "Unimplemented";  
+  return nullptr;
+}
+
+#undef DEFINE_OPERATOR
+
+
+// ---------------------- METHODS ------------------------- //
+
+// parse from a file, we will add something there to make
+// if sat --> failed (return false)
+// if unsat --> add the (assert ...)
+bool SmtlibInvariantParser::ParseInvResultFromFile(const std::string & fname) {
+  std::ifstream fin(fname);
+  if (! fin.is_open()) {
+    ILA_ERROR << "Unable to read from : " << fname;
+    return false;
+  }
+
+  std::string result;
+  if (! std::getline(fin,result) || result != std::string("unsat") ) {
+    ILA_ERROR << "The cex is not unreachable, get result:" << result;
+    return false; // unknown result, possibly failed
+  }
+
+  std::stringstream sbuf;
+  sbuf << fin.rdbuf();
+  raw_string = sbuf.str();
+  ParseSmtResultFromString("(assert " + raw_string + ")");
+  return true;
+}
+
+// parse from a string: assume we have the (assert ...) there
+void SmtlibInvariantParser::ParseSmtResultFromString(const std::string & text) {
+  auto len = text.size()+1;
+  char * buffer = new char[len];
+  strncpy(buffer,text.c_str(),len);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Easily used incorrectly; doesn't always \0-terminate or check for invalid pointers [MS-banned] (CWE-120).  ](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)